docs/design/faults-diagnosis.md

```diff
@@ -0,0 +1,83 @@
+# Training Faults Diagnosis Based on Runtime Information
+
+A training job may encounter various errors during runtime. While some errors lead 
+to immediate failure, others may not manifest right away. Presently, many systems 
+resort to resuming training on new machines to diagnose errors, significantly 
+prolonging the job's training time—especially for errors that do not immediately 
+trigger failure. For instance, a training job might hang for reasons unknown, only 
+presenting timeout errors after 30 minutes or longer.
+
+Hence, we've devised a fault diagnosis system in DLRover capable of swiftly 
+identifying and diagnosing faults upon failure or when potential failures arise.
+
+## Overview
+
+During a training job, various types of runtime data are generated, providing 
+insights into the job's status. These data typically fall into three categories: 
+training logs, CUDA runtime logs, and GPU chip metrics. Based on rules
+learned from our experiences, we can identify and diagnose training faults from those data.
+However, the availability of detailed runtime data varies depending on the 
+training framework and cluster configuration. DLRover cannot assume uniform data 
+provision across all training jobs, necessitating a flexible diagnosis system. 
+For instance, while detailed CUDA runtime logs directly indicate a training hang 
+error, the absence of such logs requires alternative approaches, such as 
+monitoring GPU chip metrics like temperature.
+
+Upon detecting a failure, the diagnosis system must pinpoint its root causes and 
+execute appropriate actions to restore the training process. For example, a brief 
+network jitter causing a network timeout may erroneously flag a machine as faulty, 
+leading to unnecessary replacement. A simple retry could suffice, avoiding the 
+heavier recovery process. To minimize recovery overhead, we propose four levels of 
+diagnosis actions, ranging from light to heavy intervention:
+1. retry, when the error is temporary
+2. restart training process
+3. re-configure training job, e.g., replace a failure machine
```

Comment:
Does `re-configure training job` means the training job launches a new node to replace a failure node?

Reply:
Actually, here I indicate any configuration change (e.g., resource or parameter configuration). And the failure nodes replacement is included.